Strengthen security considerations around validity claims not being selectively disclosable (#380)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

## Selectively-Disclosable Validity Claims {#sd-validity-claims}

An Issuer MUST NOT allow any security-critical claim to be selectively disclosable.

> The exact list of such claims will depend on the application,
> and SHOULD be listed by any application-specific profile of SD-JWT.

Reviewer: ```suggestion
The exact list of such claims will depend on the application
and SHOULD be listed by any application-specific profile of SD-JWT.
```

The following is a list of standard claim names that SHOULD be considered as
security-critical by any SD-JWT Issuer:

* `iss` (Issuer)

> * `aud` (Audience), although issuers may want to allow individual entries in the array to be selectively disclosable

Reviewer: ```suggestion
* `aud` (Audience), although issuers may want to allow individual entries in the array to be selectively disclosable
```
Uppercase MAY? Not sure...

* `exp` (Expiration Time)
* `nbf` (Not Before)
* `iat` (Issued At)

> * `jti` (JWT ID)

Reviewer 2: I am not sure about the format of this... SD-JWT is payload agnostic, so there is no guarantee the payload uses JWT claims (though it is of course a convention) - for example, W3C VCDM might use their own claims (`issuer`, `validFrom`, etc.)

Reviewer: Maybe an explicit reference to this being the default JWT claims would make this clear.

> In addition, the `cnf` (Confirmation Key) claim MUST NOT be selectively disclosable.

Reviewer 2: why? 
Issuer might be using claims other than `cnf` to do key binding?

## Selectively-Disclosable Validity Claims {#sd-validity-claims}

> An Issuer MUST NOT allow any security-critical content to be selectively disclosable.

Reviewer: Maybe we should be more specific here:
```suggestion
An Issuer MUST NOT allow any content to be selectively disclosable that is critical for evaluating the
SD-JWT's authenticity or validity.
```
(Not sure if `authenticity or validity` captures all things we might want to say here.)

Author: I think it probably captures enough.

## Selectively-Disclosable Validity Claims {#sd-validity-claims}

An Issuer MUST NOT allow any content to be selectively disclosable that is critical for evaluating the
SD-JWT's authenticity or validity.
The exact list of such content will depend on the application
and SHOULD be listed by any application-specific profile of SD-JWT.

> The following is a list of typical JWT claim names that SHOULD be considered as

Reviewer 2: ```suggestion
The following is a list of registered JWT claim names that SHOULD be considered as
```

security-critical:

* `iss` (Issuer)
* `aud` (Audience), although issuers MAY allow individual entries in the array to be selectively disclosable
* `exp` (Expiration Time)
* `nbf` (Not Before)
* `cnf` (Confirmation Key)

> Consequently, claims controlling the validity of the SD-JWT
> will typically be included in plaintext in the SD-JWT payload. Verifiers, however, cannot
> reliably depend on that and need to operate as though security-critical claims might be
> selectively disclosable. In such a case, the onus is on the Holder
> to release the relevant claims to the Verifier. And a malicious Holder may try to hide, for
> example, an expiration time (`exp`) in order to get a Verifier that "fails open"
> to accept an expired SD-JWT.

Reviewer 2: sorry, I don't fully understand the purpose of this paragraph.. 
if issuer will typically put in the plaintext, why verifier cannot rely on that?
If there is a possibility of a malicious holder, shouldn't the sentence be phrased more like a security consideration for the verifiers not to "fail open"..?

```suggestion
Consequently, Issuer will typically include claims controlling the validity of the SD-JWT
in plaintext in the SD-JWT payload. Verifiers, however, cannot
reliably depend on that and need to operate as though security-critical claims might be
selectively disclosable. In such a case, the onus is on the Holder
to release the relevant claims to the Verifier. And a malicious Holder may try to hide, for
example, an expiration time (`exp`) in order to get a Verifier that "fails open"
to accept an expired SD-JWT.
```

Author: Your suggested text reads much better and I think we can do without the malicious holder example now.
